internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher_test.go

```diff
@@ -6,74 +6,1558 @@ package githubupstreamwatcher
+				buildLogForUpdatingHostValid("minimal-idp-name", "True", "Success", `spec.githubAPI.host (\"%s\") is valid`, *validMinimalIDP.Spec.GitHubAPI.Host),
+				buildLogForUpdatingTLSConfigurationValid("minimal-idp-name", "True", "Success", "spec.githubAPI.tls.certificateAuthorityData is valid"),
+				buildLogForUpdatingGitHubConnectionValid("minimal-idp-name", "True", "Success", `spec.githubAPI.host (\"%s\") is reachable and TLS verification succeeds`, *validMinimalIDP.Spec.GitHubAPI.Host),
+				buildLogForUpdatingPhase("minimal-idp-name", "Error"),
+			},
+		},
+		{
+			name: "Client Secret error - additional data",
+			secrets: []runtime.Object{
+				func() runtime.Object {
+					badSecret := goodSecret.DeepCopy()
+					badSecret.Data["foo"] = []byte("bar")
+					return badSecret
+				}(),
+			},
+			githubIdentityProviders: []runtime.Object{validMinimalIDP},
+			wantResultingUpstreams: []v1alpha1.GitHubIdentityProvider{
+				{
+					ObjectMeta: validMinimalIDP.ObjectMeta,
+					Spec:       validMinimalIDP.Spec,
+					Status: v1alpha1.GitHubIdentityProviderStatus{
+						Phase: v1alpha1.GitHubPhaseError,
+						Conditions: []metav1.Condition{
+							buildClientCredentialsObtainedFalse(
+								t,
+								"extra keys found",
+								validMinimalIDP.Spec.Client.SecretName,
+								validMinimalIDP.Namespace,
+								upstreamwatchers.ReasonNotFound,
+							),
+							buildGitHubConnectionValidTrue(t, *validMinimalIDP.Spec.GitHubAPI.Host),
+							buildHostValidTrue(t, *validMinimalIDP.Spec.GitHubAPI.Host),
+							buildOrganizationsPolicyValidTrue(t, *validMinimalIDP.Spec.AllowAuthentication.Organizations.Policy),
+							buildTLSConfigurationValidTrue(t),
+						},
+					},
+				},
+			},
+			wantLogs: []string{
+				buildLogForUpdatingClientCredentialsObtained("minimal-idp-name", "False", "SecretNotFound", `extra keys found: secret from spec.client.SecretName (\"some-secret-name\") must be found in namespace \"some-namespace\" with type \"secrets.pinniped.dev/github-client\" and keys \"clientID\" and \"clientSecret\"`),
+				buildLogForUpdatingOrganizationPolicyValid("minimal-idp-name", "True", "Success", fmt.Sprintf(`spec.allowAuthentication.organizations.policy (\"%s\") is valid`, string(*validMinimalIDP.Spec.AllowAuthentication.Organizations.Policy))),
+				buildLogForUpdatingHostValid("minimal-idp-name", "True", "Success", `spec.githubAPI.host (\"%s\") is valid`, *validFilledOutIDP.Spec.GitHubAPI.Host),
+				buildLogForUpdatingTLSConfigurationValid("minimal-idp-name", "True", "Success", "spec.githubAPI.tls.certificateAuthorityData is valid"),
+				buildLogForUpdatingGitHubConnectionValid("minimal-idp-name", "True", "Success", `spec.githubAPI.host (\"%s\") is reachable and TLS verification succeeds`, *validFilledOutIDP.Spec.GitHubAPI.Host),
+				buildLogForUpdatingPhase("minimal-idp-name", "Error"),
+			},
```

Comment:
One more at the end, since `update status` is the last thing the controller does:

```golang
{
   name: "When update status request fails the controller will resync",
}
```

Reply:
I'm not sure how to fake out these K8s requests so that they return errors.

Reply:
This isn't too bad.  See this example:

https://github.com/vmware-tanzu/pinniped/blob/main/internal/controller/authenticator/jwtcachefiller/jwtcachefiller_test.go#L1618